src/Maybe.js

```diff
@@ -1,45 +1,82 @@
-module.exports = Maybe;
+var util = require('./internal/util.js');
 
 function Maybe(x) {
-    if (!(this instanceof Maybe)) {
-        return new Maybe(x);
-    }
+    throw new Error('Maybe cannot be directly instantiated.');
+};
+
+function _Just(x) {
     this.value = x;
+}
+util.extend(_Just, Maybe);
+
+function _Nothing() {}
+util.extend(_Nothing, Maybe);
+
+var _nothing = new _Nothing();
+
+Maybe.Nothing = function() {
+    return _nothing;
 };
 
-Maybe.of = function(x) {
-    return new Maybe(x);
+Maybe.Just = function(x) {
+    return new _Just(x);
 };
 
+Maybe.from = function(x) {
+    return x == null ? _nothing : Maybe.Just(x);
+};
+
+Maybe.of = Maybe.Just;
+
+Maybe.prototype.of = Maybe.Just;
+
+
 // functor
-Maybe.prototype.map = function(f) {
-    return this.value == null ? this : new Maybe(f(this.value));
+_Just.prototype.map = function(f) {
+    return this.of(f(this.value));
 };
 
+_Nothing.prototype.map = util.returnThis;
+
 // apply
 // takes a Maybe that wraps a function (`app`) and applies its `map`
 // method to this Maybe's value, which must be a function.
-Maybe.prototype.ap = function(m) {
-    return typeof this.value !== 'function' ? new Maybe(null) : m.map(this.value);
+_Just.prototype.ap = function(m) {
+    return m.map(this.value);
 };
 
+_Nothing.prototype.ap = util.identity;
+
 // applicative
-Maybe.prototype.of = Maybe.of;
+// `of` inherited from `Maybe`
+
 
 // chain
 //  f must be a function which returns a value
 //  f must return a value of the same Chain
 //  chain must return a value of the same Chain
-//
-Maybe.prototype.chain = function(f) {
-    return this.value == null ? this : f(this.value);
-};
+_Just.prototype.chain = util.baseMap;
+
+_Nothing.prototype.chain = util.returnThis;
+
+
+// 
+_Just.prototype.datatype = _Just;
+
+_Nothing.prototype.datatype = _Nothing;
 
 // monad
 // A value that implements the Monad specification must also implement the Applicative and Chain specifications.
 // see above.
 
 // equality method to enable testing
-Maybe.prototype.equals = function(that) {
-    return this.value === that.value;
+_Just.prototype.equals = function(that) {
+    return that.datatype = _Just && this.value === that.value;
```

Comment:
> You mean `===` here rather than `=`.

oops. guess we need jshint.

> Why that.datatype === _Just rather than that instanceof _Just

i just don't like `instanceof` :smile: . but it would probably make more sense in case we ever want to "extend" these types